### PR TITLE
Object reference preservation

### PR DESCRIPTION
**Pull request by fcanobrash**
By returning a new object for each dict/list we were removing any reference that other objects might have to them.

### CODE REVIEW

**Review by andrewbaxter — CHANGES_REQUESTED** [review]
Nice, good finding this!  I imagine this could have already been causing some strange issues.

I _think_ this solution might cause some other subtle changes at the moment... I can't come up with a good way to summarize this so just dumping what I'm looking at:

- `request_to_dict` does a shallow copy, so the metadata in the dict it returns and the metadata in the original request are the same.
- `process_object`/etc modifies the input to make changes so the original data changes
- `process_spider_input` does `parse_request` on the req before the callback is called

This could lead to the spider getting very slightly different data in the callback when running in autounit mode vs running normally.  I think this could be an issue for `process_spider_output` modifying the output requests as well - they'll be different when received by the next callback.

I'm not sure about a clean solution.  You might have to make an "object-reference preserving" deep copy and copy the requests before processing them in the middleware? (or copy them after request_to_dicting them if that's easier), which isn't impossible in Python luckily... maybe there's another way to do this though.

**Comment by fcanobrash** [comment]
Right, good catch!

I think just passing a `copy` of `elem` to `parse_object` in `process_spider_output` instead of `elem` itself would solve this issue.

I've been doing some tests and looks like it works as expected, not changing the original output while keeping references in the recorded data.

Does that make sense?

**Comment by andrewbaxter** [comment]
Copy is a shallow copy right?  I think this still risks changes in objects at lower levels.  I think it may need a fairly subtle custom deep copy.

**Comment by andrewbaxter** [comment]
Ah, standard `copy.deepcopy` preserves references, so I think that should be fine:
```
import copy
import datetime
import pickle


def generate():
    inner = {
        'a': datetime.datetime.now(),
    }
    a = {
        '1': inner,
        '2': inner,
    }
    return a, a


print('no copy')
a, b = generate()
b['1']['a'] = b['1']['a'].isoformat()  # Modify copy
b = pickle.loads(pickle.dumps(b))
print(a)
print(b)

print('shallow copy')
a, b = generate()
b = copy.copy(b)
b['1']['a'] = b['1']['a'].isoformat()  # Modify copy
b = pickle.loads(pickle.dumps(b))
print(a)
print(b)

print('deep copy')
a, b = generate()
b = copy.deepcopy(b)
b['1']['a'] = b['1']['a'].isoformat()  # Modify copy
b = pickle.loads(pickle.dumps(b))
print(a)
print(b)
```

```
no copy
{'1': {'a': '2019-10-29T19:40:30.974796'}, '2': {'a': '2019-10-29T19:40:30.974796'}}
{'1': {'a': '2019-10-29T19:40:30.974796'}, '2': {'a': '2019-10-29T19:40:30.974796'}}
shallow copy
{'1': {'a': '2019-10-29T19:40:30.974840'}, '2': {'a': '2019-10-29T19:40:30.974840'}}
{'1': {'a': '2019-10-29T19:40:30.974840'}, '2': {'a': '2019-10-29T19:40:30.974840'}}
deep copy
{'1': {'a': datetime.datetime(2019, 10, 29, 19, 40, 30, 974867)}, '2': {'a': datetime.datetime(2019, 10, 29, 19, 40, 30, 974867)}}
{'1': {'a': '2019-10-29T19:40:30.974867'}, '2': {'a': '2019-10-29T19:40:30.974867'}}
```

**Comment by fcanobrash** [comment]
@andrewbaxter, you're right about `deepcopy`; just updated the PR with this change.
Many thanks for the tests and the follow up on this

File: scrapy_autounit/middleware.py [hunk 1]
```diff
@@ -70,15 +72,15 @@ def process_spider_input(self, response, spider):
         filter_args = {'crawler', 'settings', 'start_urls'}
         if isinstance(spider, CrawlSpider):
             filter_args |= {'rules', '_rules'}
-        response.meta['_autounit'] = {
+        response.meta['_autounit'] = pickle.dumps({
```
Comment (fcanobrash) [inline scrapy_autounit/middleware.py:75]: This actually is a bit out of scope but not that much either.
Turns out that if the callback modified `request.meta` for instance, we were recording the modified version in the end (in `process_spider_output`).
This way we record the request just as it arrived to the callback.

**Review by andrewbaxter — APPROVED** [review]
LGTM

**Pull request merged**